warehouse/oidc/views.py

```diff
@@ -0,0 +1,115 @@
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.events.tags import EventTag
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    # `json_body` can return any valid top-level JSON type, so we have
+    # to make sure we're actually receiving a dictionary.
+    if not isinstance(body, dict):
+        return _invalid(
+            errors=[
+                {
+                    "code": "invalid-payload",
+                    "description": "payload is not a JSON dictionary",
+                }
+            ]
+        )
+
+    unverified_jwt = body.get("token")
+    if unverified_jwt is None:
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is missing"}]
+        )
+
+    if not isinstance(unverified_jwt, str):
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "token is not a string"}]
+        )
+
+    # For the time being, GitHub is our only OIDC provider.
+    # In the future, this should locate the correct service based on an
+    # identifier in the request body.
+    oidc_service = request.find_service(IOIDCProviderService, name="github")
+    provider = oidc_service.find_provider(unverified_jwt)
+    if not provider:
+        return _invalid(
+            errors=[
+                {"code": "invalid-token", "description": "malformed or invalid token"}
+            ]
+        )
+
+    # At this point, we've verified that the given JWT is valid for the given
+    # project. All we need to do is mint a new token.
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+    not_before = int(time.time())
+    expires_at = not_before + 900
+    serialized, dm = macaroon_service.create_macaroon(
+        request.domain,
+        f"OpenID token: {provider} ({not_before})",
```

Comment:
Should this include the provider name as well? Currently it's just stringifying the provider, which doesn't include the provider name (maybe that should instead?). E.g., "GitHub".

Reply:
Looking at this:

<img width="424" alt="image" src="https://user-images.githubusercontent.com/294415/199596349-b2df7976-f5a8-43bb-9f62-a465b564dbda.png">

The 'token name' line is pretty confusing here, because it's not really a name like other tokens. The timestamp is not human-readable either. Maybe this could be something like:

```
Creator: https://github.com/di/nachos-test/.github/workflows/publish.yml
```

which could even be a link!

Reply:
Yeah, I think so -- more generally this is making me realize that the current provider stringification isn't great (it's a syntax I made up).

Reply:
I think it'd be super nice if these were links -- think about how we might expose these publicly alongside a release.

Reply:
I agree! I think I'm going to refactor the `OIDCProvider` interface to require that every provider include its link, and I'll include that in the render.

Reply:
One hiccup with linking: we can probably link to the repository, but linking directly to the workflow itself might be a little tricky (since we don't actually know the repo's default branch name, or if the exact workflow ran even came from the default branch).

Reply:
...although we do have the `sha` claim, so maybe we could reuse that?

Reply:
Never mind, I can't use the `sha` claim in this context unless we include it in the set of DB-stored claims.

Reply:
We can bypass the issue with the default branch name by using something like:

https://github.com/pypi/warehouse/blob/HEAD/.github/workflows/ci.yml

I think we can punt on figuring out the exact SHA/workflow that resulted in a build until later.

Reply:
[0fb4679](https://github.com/pypi/warehouse/pull/11272/commits/0fb4679ca2394e08389e8a284910bdcb33b7f95e)